basecoin/store/src/avl/tree.rs

```diff
@@ -49,23 +49,143 @@ impl<K: Ord + AsBytes, V: Borrow<[u8]>> AvlTree<K, V> {
     /// Insert a value into the AVL tree, this operation runs in amortized O(log(n)).
     pub fn insert(&mut self, key: K, value: V) -> Option<V> {
         let node_ref = &mut self.root;
-        let mut old_value = None;
-        AvlTree::insert_rec(node_ref, key, value, &mut old_value);
-        old_value
+        AvlTree::insert_rec(node_ref, key, value)
     }
 
     /// Insert a value in the tree.
-    fn insert_rec(node_ref: &mut NodeRef<K, V>, key: K, value: V, old_value: &mut Option<V>) {
+    fn insert_rec(node_ref: &mut NodeRef<K, V>, key: K, value: V) -> Option<V> {
         if let Some(node) = node_ref {
+            let old_value = match node.key.cmp(&key) {
+                Ordering::Greater => AvlTree::insert_rec(&mut node.left, key, value),
+                Ordering::Less => AvlTree::insert_rec(&mut node.right, key, value),
+                Ordering::Equal => Some(node.set_value(value)),
+            };
+            node.update();
+            AvlTree::balance_node(node_ref);
+            old_value
+        } else {
+            *node_ref = as_node_ref(key, value);
+            None
+        }
+    }
+
+    /// Remove a value from the AVL tree, this operation runs in amortized O(log(n)).
+    pub fn remove(&mut self, key: K) -> Option<V> {
+        let node_ref = &mut self.root;
+        AvlTree::remove_rec(node_ref, key)
+    }
+
+    /// Remove a value from the tree.
+    fn remove_rec(node_ref: &mut NodeRef<K, V>, key: K) -> Option<V> {
+        let removed_value = if let Some(node) = node_ref {
             match node.key.cmp(&key) {
-                Ordering::Greater => AvlTree::insert_rec(&mut node.left, key, value, old_value),
-                Ordering::Less => AvlTree::insert_rec(&mut node.right, key, value, old_value),
-                Ordering::Equal => *old_value = Some(node.set_value(value)),
+                Ordering::Greater => AvlTree::remove_rec(&mut node.left, key),
+                Ordering::Less => AvlTree::remove_rec(&mut node.right, key),
+                Ordering::Equal => AvlTree::remove_top(node_ref).map(|node| node.value),
             }
+        } else {
+            None
+        };
+
+        if let Some(node) = node_ref {
             node.update();
             AvlTree::balance_node(node_ref);
+        }
+
+        removed_value
+    }
+
+    /// Removes the top node in the tree, if it exists.
+    fn remove_top(node_ref: &mut NodeRef<K, V>) -> NodeRef<K, V> {
+        let removed_node = if let Some(node) = node_ref {
+            if node.right.is_some() {
+                // Remove the leftmost node in the right subtree and replace the current.
+                let mut leftmost_node_ref = AvlTree::remove_leftmost(&mut node.right);
```

Comment:
Not really sure if this was a note. Note that, removed nodes are always a leaf. left and right nodes are set to `None`.

So, what I am doing here is:
1. point `leftmost_node.left` to where `node.left` points.
2. point `leftmost_node.right` to where `node.right` points. (since, I called, `node.{right,left}.take()`, `node` is now a leaf too.)
3. replace `node` with `leftmost_node` and return the old replaced value as the removed value.

Reply:
sorry it was a note,  that `remove_leftmost` had a misleading name for me in the beginning.